aiida/orm/nodes/data/cif.py

```diff
@@ -196,27 +197,46 @@ def pycifrw_from_cif(datablocks, loops=None, names=None):
 
 def parse_formula(formula):
     """
-    Parses the Hill formulae, written with spaces for separators.
+    Parses the Hill formulae. Does not need spaces as separators.
+    Works also for partial occupancies and for chemical groups enclosed in round/square/curly brackets.
+    Elements are counted and a dictionary is returned.
+    e.g.  'C[NH2]3NO3'  -->  {'C': 1, 'N': 4, 'H': 6, 'O': 3}
     """
-    import re
+
+    def chemcount_str_to_number(string):
+        if not string:
+            quantity = 1
+        else:
+            quantity = float(string)
+            if quantity.is_integer():
+                quantity = int(quantity)
+        return quantity
 
     contents = {}
-    for part in re.split(r'\s+', formula):
-        m = re.match(r'(\D+)([\.\d]+)?', part)
 
-        if m is None:
+    # split blocks with parentheses
+    for block in re.split(r'(\([^\)]*\)[^A-Z\(\[\{]*|\[[^\]]*\][^A-Z\(\[\{]*|\{[^\}]*\}[^A-Z\(\[\{]*)', formula):
```

Comment:
I see that you are deviating from a number of syntax rules that are [defined rigorously for the `_chemical_formula_*` tag](https://www.iucr.org/__data/iucr/cifdic_html/1/cif_core.dic/Cchemical_formula.html) of the CIF format. For example, the spaces between elements and element groups are required in the spec. Also, only parentheses are allowed to group elements, not curly brackets. So in a sense, you are now parsing chemical formula tags that are technically invalid CIF format. I am not sure whether we should do this.

Reply:
Yeah, this regex is quite more general than what one should (in principle) find in CIF files.
Nevertheless, I found that chemical formulas in MPDS cif files do not have spaces between elements and may use square brackets (e.g. `LiSc[SeO3]2`).

Also, I wanted this function to be able to parse formulas generated by `StructureData.get_formula()`, by pymtagen, and ase. These may not contain spaces.
In the case of partial occupancies, `StructureData.get_formula()` will return a formula with curly brackets.
I thought it would be useful to have one function that does all of this: for example I am using it to detect possible errors in the CIF parsing (I will make a PR to `aiida-codtools` with this soon). If you think it is useful, maybe it can be moved somewhere else, as it is not limited to CIF chemical formulas only. Otherwise we can specialize it, and one can write a more general parser for other applications.

Reply:
It is also true that one could avoid parsing the formula of a `StructureData` or a pymatgen structure, as he could obtain the set of elements and their quantities directly from the list of sites...

Reply:
As you stated yourself afterwards, it would be a bit of a merry-go-round to parse the string returned by `StructureData.get_formula()` when all information can be gotten directly from the class itself in other more efficient ways. I do see now, that even though the `parse_formula` is in the `cif` module, it was not explicitly marked as only parsing the CIF `_chemical_formula` tag, but merely Hill notation formulae in general. So I guess it is ok to keep this more flexible version.